src/Data/Text/Markdown.hs

```diff
@@ -0,0 +1,268 @@
+  heading _lvl = Mark.plain . Mark.strong
+  rawBlock (Mark.Format f) t = error . T.unpack $ "Unsupported raw " <> f <> " block:\n" <> t
+  referenceLinkDefinition = mempty
+  list _type _spacing = mconcat
+
+parseSyntax :: Text -> Either String Syntax
+parseSyntax t = case readTerm t of
+  Left e -> Left (T.unpack e)
+  Right Nothing -> Left "empty code"
+  Right (Just s) -> case processParsedTerm s of
+    Left e -> Left (T.unpack $ prettyTypeErrText t e)
+    Right (ProcessedTerm modul _req _reqCtx) -> Right $ void $ moduleAST modul
+
+findCode :: Document Syntax -> [Syntax]
+findCode = catMaybes . concatMap (map codeOnly . nodes) . paragraphs
+ where
+  codeOnly = \case
+    LeafCode s -> Just s
+    LeafCodeBlock _i s -> Just s
+    _l -> Nothing
+
+instance ToJSON (Paragraph Syntax) where
+  toJSON = String . toText
+
+instance ToJSON (Document Syntax) where
+  toJSON = String . toText
+
+instance FromJSON (Document Syntax) where
+  parseJSON v = parsePars v <|> parseDoc v
+   where
+    parseDoc = withText "markdown" fromTextM
+    parsePars = withArray "markdown paragraphs" $ \a -> do
+      (ts :: [Text]) <- mapM parseJSON $ toList a
+      fromTextM $ T.intercalate "\n\n" ts
+
+-- | Read Markdown document and parse&validate the code.
+--
+-- If you want only the document with code as `Text`,
+-- use the 'fromTextPure' function.
+fromTextM :: MonadFail m => Text -> m (Document Syntax)
+fromTextM = either fail pure . fromTextE
+
+fromTextE :: Text -> Either String (Document Syntax)
+fromTextE t = fromTextPure t >>= traverse parseSyntax
+
+-- | Read Markdown document without code validation.
+fromTextPure :: Text -> Either String (Document Text)
+fromTextPure t = do
+  let spec = Mark.rawAttributeSpec <> Mark.defaultSyntaxSpec <> Mark.rawAttributeSpec
+  let runSimple = left show . runIdentity
+  runSimple $ Mark.commonmarkWith spec "markdown" t
+
+--------------------------------------------------------------
+-- DIY STREAM
+--------------------------------------------------------------
+
+-- | This is the naive and easy way to get text from markdown document.
+toText :: ToStream a => a -> Text
+toText = streamToText . toStream
+
+-- | Token stream that can be easily converted to text or brick widgets.
+--
+-- TODO: #574 Code blocks should probably be handled separately.
+data StreamNode' t
+  = TextNode (Set TxtAttr) t
+  | CodeNode t
+  | RawNode String t
+  | ParagraphBreak
+  deriving (Eq, Show, Functor)
+
+type StreamNode = StreamNode' Text
+
+unStream :: StreamNode' t -> (t -> StreamNode' t, t)
+unStream = \case
+  TextNode a t -> (TextNode a, t)
+  CodeNode t -> (CodeNode, t)
+  RawNode a t -> (RawNode a, t)
+  ParagraphBreak -> error "Logic error: Paragraph break can not be unstreamed!"
+
+-- | Get chunks of nodes not exceeding length and broken at word boundary.
+--
+-- The split will end when no more nodes (then words) can fit or on 'ParagraphBreak'.
+chunksOf :: Int -> [StreamNode] -> [[StreamNode]]
```

Comment:
It is a shame we seem to be reimplementing some of the functionality of https://hackage.haskell.org/package/word%2Dwrap ...

Reply:
We are doing slightly more - we have annotated chunks of text, which we break while keeping the annotation.

Maybe this could be generalised as some kind of traversal?